### DISCUSSION TITLE
Relation field ordering vs performance

### DISCUSSION THREAD

**Discussion by ydewit** [souffle-lang/souffle#2362] [Q&A] [answered]
Does the order of relation fields affect indexing/performance like it does in relation databases?

I am assuming yes, but since I am new to Souffle and to Datalog it seemed worth asking.

**Comment by psubotic** [comment, marked as answer]
In souffle based on the access patterns we derive a minimal set of indexes
(b-trees with unique ordering). So the way you declare a relation does not
have any effect, but the accesses to a relation will be used to compute
b-tree orders. This generally has a price in terms of memory (due to
replication).

For more details see: http://www.vldb.org/pvldb/vol12/p141-subotic.pdf

Note: that the indexing and literal ordering problems cannot be fully
decoupled and the literal orders in a rule affect the indexing problem. We
have some work for offline literal orders that take into account indexing.

For more details see: https://psubotic.github.io/papers/lopstr2022.pdf

**Reply by ydewit** [reply]
Thank you!
